base/base/bit_cast.h

```diff
@@ -13,6 +14,12 @@ template <typename To, typename From>
 std::decay_t<To> bit_cast(const From & from)
 {
     To res {};
-    memcpy(static_cast<void*>(&res), &from, std::min(sizeof(res), sizeof(from)));
+    if constexpr (std::endian::native == std::endian::little || !std::is_arithmetic_v<To>)
```

Comment:
ClickHouse's `bit_cast` ideally remains as close as possible in terms of behavior to `std::bit_cast` which is doing a plain `memcpy` without reversing bytes.

To be honest, I don't really understand why we should build special handling for
- big endian or
- arithmetic types on little endian (*)

into this low-level utility. Does reversing arithmetic values (*) not also change/break the existing behavior on little endian?

Perhaps this fixes the issue but I would feel a lot easier if the fix is moved out of bit_cast into some datetime64 handling code.

Reply:
Realized no need for reverseMemcpy, offset cal should take care of the fix. 
Removed arithmetic check as suggested.
If not in bit_cast.h, we will have to update a dozen place to make sure the bit_cast in big endian machine works properly..